cli/tests/ld-floxlib.bats

```diff
@@ -0,0 +1,117 @@
+}
+
+# ---------------------------------------------------------------------------- #
+
+# Helpers for project based tests.
+
+project_setup() {
+  export PROJECT_DIR="${BATS_TEST_TMPDIR?}/project-${BATS_TEST_NUMBER?}"
+  export PROJECT_NAME="${PROJECT_DIR##*/}"
+  rm -rf "$PROJECT_DIR"
+  mkdir -p "$PROJECT_DIR"
+  cp ./ld-floxlib/* "$PROJECT_DIR"
+  pushd "$PROJECT_DIR" > /dev/null || return
+  _PKGDB_GA_REGISTRY_REF_OR_REV="${PKGDB_NIXPKGS_REV_OLDER?}" \
+    "$FLOX_BIN" init
+}
+
+project_teardown() {
+  popd > /dev/null || return
+  rm -rf "${PROJECT_DIR?}"
+  unset PROJECT_DIR
+  unset PROJECT_NAME
+}
+
+# ---------------------------------------------------------------------------- #
+
+setup() {
+  common_test_setup
+  project_setup
+}
+teardown() {
+  project_teardown
+  common_test_teardown
+}
+
+# ---------------------------------------------------------------------------- #
+#
+@test "test ld-floxlib.so on Linux only" {
+  if [ $(uname -s) != "Linux" ]; then
+    skip "not Linux"
+  fi
+
+  # Note:
+  # - installing old versions of nix (2.10.3) and glibc (2.34) for use in tests
+  # - installing curl and libarchive because those packages provide libraries
+  #   that are runtime dependencies of libnixmain.so
+  run env _PKGDB_GA_REGISTRY_REF_OR_REV="${PKGDB_NIXPKGS_REV_OLDER?}" \
+    "$FLOX_BIN" install curl gcc glibc libarchive nix patchelf
+  assert_success
+  assert_output --partial "✅ 'curl' installed to environment"
+  assert_output --partial "✅ 'gcc' installed to environment"
+  assert_output --partial "✅ 'glibc' installed to environment"
+  assert_output --partial "✅ 'libarchive' installed to environment"
+  assert_output --partial "✅ 'nix' installed to environment"
+  assert_output --partial "✅ 'patchelf' installed to environment"
+
+  ### Test 1: load libraries found in $FLOX_ENV_LIB_DIRS last
+  run "$FLOX_BIN" activate -- bash ./test-load-library-last.sh < /dev/null
+  assert_success
+
+  ### Test 2: confirm LD_AUDIT can find missing libraries
+  # Link against nixmain because that's a library that won't be present on any host system.
+  # Build print-nix-version, remove RUNPATH & interpreter
+  run "$FLOX_BIN" activate -- bash -exc '" \
+    g++ -std=c++17 -o get-nix-version ./get-nix-version.cc -L"$FLOX_ENV"/lib -lnixmain && \
+    patchelf --remove-rpath ./get-nix-version && \
+    patchelf --set-interpreter "$( \
+      patchelf --print-interpreter /bin/sh \
+    )" ./get-nix-version && \
+    LD_FLOXLIB_DEBUG=1 ./get-nix-version"'
+  assert_success
+  assert_output --partial "testing (Nix) 2.10.3"
+
+  ### Test 3: confirm binary cannot find missing libraries without LD_AUDIT
+  # Note run with "run -127" to silence the 127 "Command not found" error code
+  # warning that bats will display by default when it attempts to launch a
+  # command that fails to run because it cannot load its libraries.
+  run -127 "$FLOX_BIN" activate -- bash -exc \
+    '"env -i LD_DEBUG=libs ./get-nix-version"'
+  assert_failure
```

Comment:
> for my understanding, is running ./get-nix-version (without flox activate) supposed to work?

No it is not. It relies upon $LD_AUDIT to locate the shared libraries in $FLOX_ENV/lib.

> My understanding is that once the outlink to the env is deleted, this will cease to work.

Correct. This whole approach is only expected to work within the spawned development shell. It's not like a Nix packaged application that forces the use of Nix packaged prerequisites only. We need to not get confused by this; the purpose of this patch is to avoid breaking non-Nix ELF binaries invoked from within flox-activated environments. For those Nix artifacts created within these environments, we will need to rely on wrapping and other Nix tricks to ensure they find what they need.

Reply:
I.e. any binary i build while within a flox environment, only works if i'm in the same environment or another with the same libs?

Reply:
_Maybe?_ The Nix-provided gcc-wrapper will produce ELF binaries:
* configured with a Nix-provided dynamic linker interpreter that is hard-coded to look for `glibc` from its own package, so you'll find that binaries built in your flox env will be able to find the "right" Nix-provided glibc unless you apply the LD_LIBRARY_PATH hammer
* configured with a RUNPATH that looks for the necessary libraries in each of the individual /nix/store paths:
<pre><font color="#5F5FAF"><b>flox </b></font><font color="#FFAF87"><b>[tmp.vYoEM7vY0o]</b></font> <font color="#26A269">brantley@fedora</font>:<font color="#26A269">/tmp/tmp.vYoEM7vY0o</font>$ cc -o get-glibc-version ./get-glibc-version.c -I&quot;$FLOX_ENV&quot;/include -L&quot;$FLOX_ENV&quot;/lib
<font color="#5F5FAF"><b>flox </b></font><font color="#FFAF87"><b>[tmp.vYoEM7vY0o]</b></font> <font color="#26A269">brantley@fedora</font>:<font color="#26A269">/tmp/tmp.vYoEM7vY0o</font>$ ldd ./get-glibc-version
	linux-vdso.so.1 (0x0000ffff82e86000)
	libc.so.6 =&gt; /nix/store/l0y717pj7ds1agfpzrlz4zq112hnpjqq-glibc-2.34-210/lib/libc.so.6 (0x0000ffff82cb7000)
	/nix/store/l0y717pj7ds1agfpzrlz4zq112hnpjqq-glibc-2.34-210/lib/ld-linux-aarch64.so.1 (0x0000ffff82e53000)
<font color="#5F5FAF"><b>flox </b></font><font color="#FFAF87"><b>[tmp.vYoEM7vY0o]</b></font> <font color="#26A269">brantley@fedora</font>:<font color="#26A269">/tmp/tmp.vYoEM7vY0o</font>$ objdump -p get-glibc-version | grep PATH
  RUN<font color="#C01C28"><b>PATH</b></font>              /nix/store/l0y717pj7ds1agfpzrlz4zq112hnpjqq-glibc-2.34-210/lib:/nix/store/hmpj6gqijarx6iq1rzmf02sql6wbb58j-gcc-9.5.0-lib/lib
<font color="#5F5FAF"><b>flox </b></font><font color="#FFAF87"><b>[tmp.vYoEM7vY0o]</b></font> <font color="#26A269">brantley@fedora</font>:<font color="#26A269">/tmp/tmp.vYoEM7vY0o</font>$ 
</pre>

... so you'll find that things built in flox environments can find their dependencies if they're there.

Reply:
Okk so unless you nuke the nix store (or go it) it should still work 👍🏼